core/trino-main/src/main/java/io/trino/execution/SqlStage.java

```diff
@@ -268,16 +268,22 @@ public void recordGetSplitTime(long start)
         stateMachine.recordGetSplitTime(start);
     }
 
-    private synchronized void updateTaskStatus(TaskStatus status)
+    private void updateTaskStatus(TaskStatus status)
     {
-        if (status.getState().isDone()) {
-            finishedTasks.add(status.getTaskId());
+        boolean isDone = status.getState().isDone();
+        if (!isDone && stateMachine.getState() == StageState.RUNNING) {
+            return;
         }
-        if (!finishedTasks.containsAll(allTasks)) {
-            stateMachine.transitionToRunning();
-        }
-        else {
-            stateMachine.transitionToPending();
+        synchronized (this) {
+            if (isDone) {
+                finishedTasks.add(status.getTaskId());
+            }
+            if (finishedTasks.size() == allTasks.size()) {
```

Comment:
I don't know if this is a valid change with fault tolerant execution.
If `allTasks` may contain failed task attempts, then this could be wrong.
@arhimondr could you PTAL at this one ?

Reply:
Without this change, `finishedTasks` need to contain all tasks from `allTasks`. Failed, canceled and aborted tasks should end up in `finishedTasks`.

Reply:
yes, maybe this is correct then, will wait for @arhimondr to confirm though.

Reply:
Yeah, it should contain all for a stage to produce a final stage info (including both failed and successful retries)